[GStreamer][WebRTC] replaceTrack support

philn opened the pull request:
#### 5bd55e4bf534024c800d3823cde87ba2139cb3ff
<pre>
[GStreamer][WebRTC] replaceTrack support
<a href="https://bugs.webkit.org/show_bug.cgi?id=252820">https://bugs.webkit.org/show_bug.cgi?id=252820</a>

Reviewed by Xabier Rodriguez-Calvar.

In the GStreamer WebRTC backend `sender.replaceTrack()` is now supported. Under the hood, the
outgoing media source now has an input-selector just after the mediastreamsrc element. This allows
us to toggle between a mediastreamsrc and a fake source, eg, when replacing the track with null.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/webrtc/video-replace-track.html:
* Source/WebCore/Modules/mediastream/RTCRtpTransceiverBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::setConfiguration):
(WebCore::GStreamerMediaEndpoint::configureAndLinkSource):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::addTrack):
(WebCore::GStreamerMediaEndpoint::removeTrack):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::createLinkedSourceForTrack):
(WebCore::GStreamerMediaEndpoint::createDataChannel):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
(WebCore::GStreamerMediaEndpoint::setSenderSourceFromTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::createLinkedSourceForTrack):
(WebCore::GStreamerPeerConnectionBackend::setSenderSourceFromTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::ensureDebugCategoryIsRegistered):
(WebCore::GStreamerRtpSenderBackend::GStreamerRtpSenderBackend):
(WebCore::m_initData):
(WebCore::GStreamerRtpSenderBackend::startSource):
(WebCore::GStreamerRtpSenderBackend::stopSource):
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:
(WebCore::GStreamerRtpTransceiverBackend::stop):
(WebCore::GStreamerRtpTransceiverBackend::stopped const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkOutgoingSource):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::flush):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::isStopped const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkOutgoingSource):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkOutgoingSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::flush):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/260815@main">https://commits.webkit.org/260815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2b66428de5078ab2b4a262844aaf2b357248b15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1001 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118660 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113388 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20094 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9825 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101780 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115259 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98196 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43172 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29849 "Found 1 new test failure: imported/w3c/web-platform-tests/preload/onerror-event.html (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84936 "Found 2 new API test failures: /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11366 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31192 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12029 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8123 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17395 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50796 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7493 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13763 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/RTCRtpTransceiverBackend.h:
```diff
@@ -26,6 +26,7 @@
 
 #if ENABLE(WEB_RTC)
 
+#include "ExceptionOr.h"
```
Do you need this here?

philn replied:
Somehow this PR disaligns the stars and without this include:

```
                                                                                          │
│In file included from /app/webkit/Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:21:                                                                                            │
│In file included from /app/webkit/Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.h:27:                                                                                              │
│/app/webkit/Source/WebCore/Modules/mediastream/RTCRtpTransceiverBackend.h:46:13: error: no template named 'ExceptionOr'                                                                                          │
│    virtual ExceptionOr<void> setCodecPreferences(const Vector<RTCRtpCodecCapability>&) = 0;                                                                                                                     │
│            ^                                                                                                                                                                                                    │
│In file included from /app/webkit/Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:21:                                                                                            │
│/app/webkit/Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.h:50:5: error: no template named 'ExceptionOr'                                                                           │
│    ExceptionOr<void> setCodecPreferences(const Vector<RTCRtpCodecCapability>&) final;                                                                                                                           │
│    ^                                                                                                                                                                                                            │
│2 errors generated.                                                                                                        
```

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -274,8 +274,9 @@ bool GStreamerMediaEndpoint::setConfiguration(MediaEndpointConfiguration& config
-    gst_element_set_state(m_pipeline.get(), GST_STATE_PLAYING);
+    gst_element_set_state(m_pipeline.get(), GST_STATE_READY);
     GST_DEBUG_OBJECT(m_pipeline.get(), "End-point ready");
+    gst_element_set_state(m_pipeline.get(), GST_STATE_PLAYING);
```
Why do you need these two state changes so close to each other that might be issued almost simultaneusly from multithread perspective?

philn replied:
Logging purposes.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -575,8 +576,31 @@ void GStreamerMediaEndpoint::processSDPMessage(const GstSDPMessage* message, Fun
 }
 
-void GStreamerMediaEndpoint::configureAndLinkSource(RealtimeOutgoingMediaSourceGStreamer& source)
+void GStreamerMediaEndpoint::configureAndLinkSource(RealtimeOutgoingMediaSourceGStreamer& source, bool lookForUnusedPads)
```
Nit: a better name could be `shouldLookForUnusedPads`

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -575,8 +576,31 @@ void GStreamerMediaEndpoint::processSDPMessage(const GstSDPMessage* message, Fun
+
+            GRefPtr<GstCaps> codecPreferences;
+            g_object_get(transceiver.get(), "codec-preferences", &codecPreferences.outPtr(), nullptr);
+            if (gst_caps_is_always_compatible(codecPreferences.get(), allowedCaps.get())) {
```
Can't `codecPreferences` be null at this point? Should we protect about it as we do for `transceiver`?

philn replied:
i guess!

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -594,30 +618,41 @@ void GStreamerMediaEndpoint::configureAndLinkSource(RealtimeOutgoingMediaSourceG
-    if (!sinkPad) {
+    auto requestPad = [&](const String& padId) -> GRefPtr<GstPad> {
+        if (padId == "sink_%u"_s)
+            return gst_element_request_pad_simple(m_webrtcBin.get(), padId.ascii().data());
```
Important: You need to adopt here or you're leaking. It's transfer full and if you don't specify how construct the smart ptr it will use the default constructor that will increase the ref.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -594,30 +618,41 @@ void GStreamerMediaEndpoint::configureAndLinkSource(RealtimeOutgoingMediaSourceG
-        sinkPad = adoptGRef(gst_element_request_pad(m_webrtcBin.get(), padTemplate, padId.utf8().data(), caps.get()));
-    }
+        auto sinkPad = adoptGRef(gst_element_request_pad(m_webrtcBin.get(), padTemplate, padId.ascii().data(), caps.get()));
+        return sinkPad;
```
Nit: You can collapse these two lines.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -594,30 +618,41 @@ void GStreamerMediaEndpoint::configureAndLinkSource(RealtimeOutgoingMediaSourceG
 }
 
-GRefPtr<GstPad> GStreamerMediaEndpoint::requestPad(unsigned mlineIndex, const GRefPtr<GstCaps>& allowedCaps, const String& mediaStreamID)
+GRefPtr<GstPad> GStreamerMediaEndpoint::requestPad(std::optional<unsigned> mlineIndex, const GRefPtr<GstCaps>& allowedCaps, const String& mediaStreamID)
```
Capitalization of the parameter should be `mLineIndex`.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -959,15 +995,32 @@ GStreamerRtpSenderBackend::Source GStreamerMediaEndpoint::createSourceForTrack(M
+            source->start();
+        });
+    }, [&](std::nullptr_t&) {
+        RELEASE_ASSERT_NOT_REACHED();
```
Do you want to ASSERT on release mode here? Maybe just an assert?

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
```diff
@@ -78,7 +78,7 @@ class GStreamerMediaEndpoint : public ThreadSafeRefCounted<GStreamerMediaEndpoin
     bool isNegotiationNeeded(uint32_t eventId) const { return eventId == m_negotiationNeededEventId; }
 
-    void configureAndLinkSource(RealtimeOutgoingMediaSourceGStreamer&);
+    void configureAndLinkSource(RealtimeOutgoingMediaSourceGStreamer&, bool lookForUnusedPads = false);
```
`should...`

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
```diff
@@ -114,31 +115,75 @@ bool RealtimeOutgoingAudioSourceGStreamer::setPayloadType(const GRefPtr<GstCaps>
+void RealtimeOutgoingAudioSourceGStreamer::connectFallbackSource()
+{
+    if (!m_fallbackPad) {
+        m_fallbackSource = gst_element_factory_make("audiotestsrc", nullptr);
```
`makeGStreamerElement` ?

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
```diff
@@ -42,11 +42,13 @@ RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer(const
 {
     m_bin = gst_bin_new(nullptr);
 
+    m_inputSelector = gst_element_factory_make("input-selector", nullptr);
```
`makeGStreamerElement`?

philn replied:
No, this is a core element, I thought we agreed to use makeGStreamerElement for non-core elements only :)

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
```diff
@@ -143,28 +143,86 @@ bool RealtimeOutgoingVideoSourceGStreamer::setPayloadType(const GRefPtr<GstCaps>
+{
+    GST_DEBUG_OBJECT(m_bin.get(), "Connecting fallback video source");
+    if (!m_fallbackPad) {
+        m_fallbackSource = gst_element_factory_make("videotestsrc", nullptr);
```
mGE

calvaris reviewed (approved):
There are some important change you need to do and some other nits, but other than that, you're good to go.

philn committed:
[GStreamer][WebRTC] replaceTrack support
https://bugs.webkit.org/show_bug.cgi?id=252820

Reviewed by Xabier Rodriguez-Calvar.

In the GStreamer WebRTC backend `sender.replaceTrack()` is now supported. Under the hood, the
outgoing media source now has an input-selector just after the mediastreamsrc element. This allows
us to toggle between a mediastreamsrc and a fake source, eg, when replacing the track with null.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/webrtc/video-replace-track.html:
* Source/WebCore/Modules/mediastream/RTCRtpTransceiverBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::setConfiguration):
(WebCore::GStreamerMediaEndpoint::configureAndLinkSource):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::addTrack):
(WebCore::GStreamerMediaEndpoint::removeTrack):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::createLinkedSourceForTrack):
(WebCore::GStreamerMediaEndpoint::createDataChannel):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
(WebCore::GStreamerMediaEndpoint::setSenderSourceFromTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::createLinkedSourceForTrack):
(WebCore::GStreamerPeerConnectionBackend::setSenderSourceFromTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::ensureDebugCategoryIsRegistered):
(WebCore::GStreamerRtpSenderBackend::GStreamerRtpSenderBackend):
(WebCore::m_initData):
(WebCore::GStreamerRtpSenderBackend::startSource):
(WebCore::GStreamerRtpSenderBackend::stopSource):
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:
(WebCore::GStreamerRtpTransceiverBackend::stop):
(WebCore::GStreamerRtpTransceiverBackend::stopped const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkOutgoingSource):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::flush):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::isStopped const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkOutgoingSource):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::connectFallbackSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::unlinkOutgoingSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkOutgoingSource):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::flush):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: https://commits.webkit.org/260815@main
